src/Build/Instance/TaskRegistry.cs

```diff
@@ -1666,34 +1666,41 @@ internal bool TaskBodyEvaluated
                         // Cannot have a null or empty name for the type after expansion.
                         ProjectErrorUtilities.VerifyThrowInvalidProject
                         (
-                        !String.IsNullOrEmpty(expandedType),
-                        parameter.ParameterTypeLocation,
-                        "InvalidEvaluatedAttributeValue",
-                        expandedType,
-                        parameter.ParameterType,
-                        XMakeAttributes.parameterType,
-                        XMakeElements.usingTaskParameter
+                            !String.IsNullOrEmpty(expandedType),
+                            parameter.ParameterTypeLocation,
+                            "InvalidEvaluatedAttributeValue",
+                            expandedType,
+                            parameter.ParameterType,
+                            XMakeAttributes.parameterType,
+                            XMakeElements.usingTaskParameter
                         );
 
-                        // Try and get the type directly 
-                        Type paramType = Type.GetType(expandedType);
-
-                        // The type could not be got directly try and see if the type can be found by appending the FrameworkAssemblyName to it.
-                        if (paramType == null)
+                        Type paramType;
+                        if (expandedType.StartsWith("Microsoft.Build.Framework.", StringComparison.OrdinalIgnoreCase) && !expandedType.Contains(","))
                         {
-                            paramType = Type.GetType(expandedType + "," + typeof(ITaskItem).GetTypeInfo().Assembly.FullName, false /* don't throw on error */, true /* case-insensitive */);
+                            // This is workaround for internal bug https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1448821
+                            // Visual Studio can load different version of Microsoft.Build.Framework.dll and non fully classified type could be resolved from it 
+                            // which cause InvalidProjectFileException with "UnsupportedTaskParameterTypeError" message.
+                            // Another way to address this is to load types from compiled assembly - that would be more robust solution but also much more complex and risky code changes.
+                            paramType = Type.GetType(expandedType + "," + typeof(ITaskItem).GetTypeInfo().Assembly.FullName, false /* don't throw on error */, true /* case-insensitive */) ??
+                                        Type.GetType(expandedType);
+                        }
+                        else
+                        {
+                            paramType = Type.GetType(expandedType) ??
+                                        Type.GetType(expandedType + "," + typeof(ITaskItem).GetTypeInfo().Assembly.FullName, false /* don't throw on error */, true /* case-insensitive */);
```

Comment:
Is this really the only source of an ITaskItem? It's public, so I imagine someone has probably extended it at this point?

Reply:
I am not sure if I follow. What is the concern in particular?
`ITaskItem` is interface and we own it, so it is under our control if its definition will always be inside Framework assembly - unless we chose otherwise.

Reply:
I'm wondering about `MyCustomTaskItem : ITaskItem`. MyCustomTaskItem is defined in MyCustomAssembly, but I think it's a valid task input, right?

Reply:
It is sorta-kinda-just-barely possible to pass a custom `ITaskItem` back from a task and thread it through to another one, but I don't think that's relevant here, because this is about specifying the input types to a code task. I don't think it's legal to restrict an input property to `MyCustomTaskItem`, because MSBuild itself can't synthesize one. If you did have such a custom thing, you must either fully qualify the type name (so the first clause here would work) or have already made it available (so the first clause would work).